src/ucp/api/ucp_def.h

```diff
@@ -576,6 +604,62 @@ typedef ucs_status_t (*ucp_am_callback_t)(void *arg, void *data, size_t length,
 
 
+/**
+ * @ingroup UCP_ENDPOINT
+ * @brief Callback to process incoming Active Message sent by
+ * @ref ucp_am_send_nbx routine.
+ *
+ * The callback is always called from the progress context, therefore calling
+ * communication progress routines from this callback is not allowed. It is
+ * recommended to define callbacks with relatively short execution time to
+ * avoid blocking of communication progress.
+ *
+ * @param [in]  arg           User-defined argument.
+ * @param [in]  header        User defined active message header. Can be NULL.
+ * @param [in]  header_length Active message header length in bytes.
+ * @param [in]  data          Points to the received data if @a
+ *                            UCP_AM_RECV_ATTR_FLAG_RNDV flag is not set in
+ *                            @ref ucp_am_recv_param_t.recv_attr. Otherwise
+ *                            it points to the internal UCP descriptor which
+ *                            can further be used for initiating data receive
+ *                            by using @ref ucp_am_data_recv_nbx routine.
+ * @param [in]  length        Length of data. If @a UCP_AM_RECV_ATTR_FLAG_RNDV
+ *                            flag is set in @ref ucp_am_recv_param_t.recv_attr,
+ *                            it indicates the required receive buffer size for
+ *                            initiating rendezvous protocol.
+ * @param [in]  param         Data receive parameters.
+ *
+ * @return UCS_OK         @a data will not persist after the callback returns.
+ *                        If UCP_AM_RECV_ATTR_FLAG_RNDV flag is set in
+ *                        @a param->recv_attr, data descriptor will be dropped
+ *                        and the corresponding @ref ucp_am_send_nbx call should
+ *                        complete with UCS_OK status.
+ *
+ * @return UCS_INPROGRESS Can only be returned if @a param->recv_attr flags
+ *                        contains UCP_AM_RECV_ATTR_FLAG_DATA or
+ *                        UCP_AM_RECV_ATTR_FLAG_RNDV. The @a data will persist
+ *                        after the callback has returned. To free the memory,
+ *                        a pointer to the data must be passed into
+ *                        @ref ucp_am_data_release or, in case of rendezvous
+ *                        descriptor, data receive is initiated by
+ *                        @ref ucp_am_data_recv_nbx.
+ *
+ * @return otherwise      Can only be returned if @a param->recv_attr contains
+ *                        UCP_AM_RECV_ATTR_FLAG_RNDV. In this case data
+ *                        descriptor @a data will be dropped and the
+ *                        corresponding @ref ucp_am_send_nbx call should
+ *                        complete with the status returned from the callback.
+ *
+ * @note This callback should be set and released
+ *       by @ref ucp_worker_set_am_data_handler function.
+ *
+ */
+typedef ucs_status_t (*ucp_am_data_callback_t)(void *arg, const void *header,
```

Comment:
ucp_am_recv_callback_t

Reply:
why is it more descriptive than `data`? Data is consistent with existing UCP and UCT APIs

Reply:
because we always receive the active message, but not always call recv-data
we have a flag called DATA which indicates if we can recv-data or not